server/src-lib/Hasura/GraphQL/Transport/HTTP.hs

```diff
@@ -150,6 +183,60 @@ runGQBatched env logger reqId responseErrorsConfig userInfo ipAddress reqHdrs qu
     try = flip catchError (pure . Left) . fmap Right
 
 
+runQueryDB
+  :: ( MonadIO m
+     , MonadError QErr m
+     , MonadReader E.ExecutionCtx m
+     , MonadQueryLog m
+     , MonadTrace m
+     , MonadExecuteQuery m
+     )
+  => RequestId
+  -> (GQLReqUnparsed, GQLReqParsed)
+  -> [QueryRootField UnpreparedValue]
+  -> UserInfo
+  -> (Tracing.TraceT (LazyTx QErr) EncJSON, EQ.GeneratedSqlMap)
+  -> m (DiffTime, Telem.QueryType, EncJSON)
+  -- ^ Also return 'Mutation' when the operation was a mutation, and the time
+  -- spent in the PG query; for telemetry.
+runQueryDB reqId (query, queryParsed) asts _userInfo (tx, genSql) =  do
+  -- log the generated SQL and the graphql query
+  E.ExecutionCtx logger _ pgExecCtx _ _ _ _ <- ask
+  logQueryLog logger query (Just genSql) reqId
+  (telemTimeIO, respE) <- withElapsedTime $ runExceptT $ trace "pg" $
+    Tracing.interpTraceT id $ executeQuery queryParsed asts (Just genSql) pgExecCtx Q.ReadOnly tx
+  resp <- liftEither respE
+  let !json = encodeGQResp $ GQSuccess $ encJToLBS $ snd resp
```

Comment:
Returned HTTP headers get discarded here, but they need to be sent to the client, please.

Reply:
Done!